File: app/src/main/java/pl/tysia/maggstone/ui/presentation_logic/adapter/SendingCatalogAdapter.java
package pl.tysia.maggstone.ui.presentation_logic.adapter;


import android.app.Application;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

import pl.tysia.maggstone.R;
import pl.tysia.maggstone.data.service.QueueItem;

public class SendingCatalogAdapter extends CatalogAdapter<QueueItem, SendingCatalogAdapter.SendingViewHolder> {

    public SendingCatalogAdapter(ArrayList<QueueItem> items) {
        super(items);

    }

    public class SendingViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView title;
        ProgressBar progress;

        private Observer<Integer> observer;

        SendingViewHolder(View v) {
            super(v);

            title = v.findViewById(R.id.name_tv);
            progress = v.findViewById(R.id.progress_bar);

            observer = new Observer<Integer>() {
                @Override
                public void onChanged(Integer integer) {
                    progress.setProgress(integer);
                }
            };

            v.setOnClickListener(this);

        }

        public void setObserver() {
            if (getAdapterPosition() >= 0)
                allItems.get(getAdapterPosition()).getPercentSent().observeForever(observer);
        }

        public void removeObserver(){
            if (getAdapterPosition() >= 0)
                allItems.get(getAdapterPosition()).getPercentSent().removeObserver(observer);
        }

        @Override
        public void onClick(View v) {
            int pos = getAdapterPosition();
            onItemClick(v, pos);
        }

    }

    @NonNull
    @Override
    public SendingViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View v = LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.list_item_sending_state, viewGroup, false);

        v.setLayoutParams(new RecyclerView.LayoutParams(RecyclerView.LayoutParams.MATCH_PARENT, RecyclerView.LayoutParams.WRAP_CONTENT));

        SendingViewHolder vh = new SendingViewHolder(v);
        return vh;
    }

    @Override
    public void onBindViewHolder(@NonNull SendingViewHolder catalogItemViewHolder, int i) {
        QueueItem item = shownItems.get(i);

        SendingViewHolder wareViewHolder = (SendingViewHolder) catalogItemViewHolder;

        wareViewHolder.title.setText(item.getTitle());
        wareViewHolder.progress.setProgress(item.getPercentSent().getValue());

        wareViewHolder.setObserver();

    }


    @Override
    public void onViewDetachedFromWindow(@NonNull SendingViewHolder holder) {
        super.onViewDetachedFromWindow(holder);

        holder.removeObserver();

    }
}

File: app/src/main/java/pl/tysia/maggstone/ui/presentation_logic/adapter/SendingCatalogAdapter.kt
package pl.tysia.maggstone.ui.presentation_logic.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.RecyclerView
import pl.tysia.maggstone.R
import pl.tysia.maggstone.data.service.QueueItem
import java.util.*

class SendingCatalogAdapter(items: ArrayList<QueueItem>) :
    CatalogAdapter<QueueItem, SendingCatalogAdapter.SendingViewHolder>(
        items
    ) {

    inner class SendingViewHolder internal constructor(v: View) :
        RecyclerView.ViewHolder(v), View.OnClickListener {

        var title: TextView = v.findViewById(R.id.name_tv)
        var progress: ProgressBar = v.findViewById(R.id.progress_bar)

        private val observer: Observer<Int>

        fun setObserver() {
            if (adapterPosition >= 0) allItems[adapterPosition]!!.percentSent.observeForever(
                observer
            )
        }

        fun removeObserver() {
            if (adapterPosition >= 0) allItems[adapterPosition]!!.percentSent.removeObserver(
                observer
            )
        }

        override fun onClick(v: View) {
            val pos = adapterPosition
            onItemClick(v, pos)
        }

        init {
            observer = Observer { integer -> progress.progress = integer!! }
            v.setOnClickListener(this)
        }
    }

    override fun onCreateViewHolder(
        viewGroup: ViewGroup,
        i: Int
    ): SendingViewHolder {
        val v = LayoutInflater.from(viewGroup.context)
            .inflate(R.layout.list_item_sending_state, viewGroup, false)
        v.layoutParams = RecyclerView.LayoutParams(
            RecyclerView.LayoutParams.MATCH_PARENT,
            RecyclerView.LayoutParams.WRAP_CONTENT
        )
        return SendingViewHolder(
            v
        )
    }

    override fun onBindViewHolder(
        catalogItemViewHolder: SendingViewHolder,
        i: Int
    ) {
        val item = shownItems[i]

        catalogItemViewHolder.title.text = item.getTitle()
        catalogItemViewHolder.progress.progress = item.percentSent.value!!
        catalogItemViewHolder.setObserver()
    }

    override fun onViewDetachedFromWindow(holder: SendingViewHolder) {
        super.onViewDetachedFromWindow(holder)
        holder.removeObserver()
    }
}